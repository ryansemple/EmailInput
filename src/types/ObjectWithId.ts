export interface ObjectWithId {
	id: number | string
};

export default ObjectWithId;