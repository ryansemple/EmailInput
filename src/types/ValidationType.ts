export enum ValidationType {
	Success,
	Error
}