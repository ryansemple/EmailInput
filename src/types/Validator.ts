interface Validator {
	validationPredicateToPass: (email: string) => boolean
	errorMessageIfFailed: string;
}

export default Validator;