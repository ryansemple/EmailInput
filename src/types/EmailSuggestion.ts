import ObjectWithId from "./ObjectWithId";
import { returnNewUuid } from "../utility/Uuid";
import { atCharacter } from "../utility/String";

interface EmailSuggestionInterface extends ObjectWithId {
	id: string,
	domain: string,
	extension: string,
	returnFullEmail: (emailUsername: string) => string,
	returnEmailAfterAtCharacter: () => string
}

/**
 * Class that restricts the way an EmailSuggestion can be created
 * by encapsulating the logic for creating a new id so that it will
 * always be a new uuid and making the properties readonly so that
 * they can not be modified after the class has been constructed.
 *
 * @class EmailSuggestion
 * @implements {EmailSuggestionInterface}
 */
class EmailSuggestion implements EmailSuggestionInterface {

	/**
	 * The id that will be used to uniquely identity the EmailSuggestion.
	 */
	public readonly id: string;

	/**
	 * the domain part of the email. 
	 * 
	 * For example: "gmail", "yahoo".
	 */
	public readonly domain: string;

	/**
	 * The extension part of the email.
	 * 
	 * For example: "com", "ca";
	 */
	public readonly extension: string;

	constructor(domain: string, extension: string)
	{ 
		this.id = returnNewUuid();
		this.domain = domain;
		this.extension = extension;
	}

	/**
	 * Returns the full email of based on the email passed in. 
	 *
	 * For example, if "jason@g" is passed in and the emails domain 
	 * is "gmail" and the extension is "ca" then it
	 * would output "jason@gmail.ca".
	*/
	public returnFullEmail = (enteredEmail: string): string => 
	{
		let emailUsername = enteredEmail.split(atCharacter)[0];
		return `${emailUsername}@${this.returnEmailAfterAtCharacter()}`;
	}

	/**
	 * Returns the domain and the extension as a string with a 
	 * period in between them.
	 *
	 * For example, if the domain is "gmail" and the extension is 
	 * "ca" then it would output "gmail.ca".
	*/
	public returnEmailAfterAtCharacter = (): string =>
	{
		return `${this.domain}.${this.extension}`;
	}
}

export default EmailSuggestion;