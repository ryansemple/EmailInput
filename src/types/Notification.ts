import ObjectWithId from "../types/ObjectWithId";

export enum NotificationType {
	Success,
	Error
}

export interface NotificationInstance extends ObjectWithId {
	text: string,
	id: string,
	notificationType: NotificationType
}