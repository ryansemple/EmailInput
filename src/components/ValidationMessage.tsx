import React from "react";
import { ValidationType } from "../types/ValidationType";
import { red, green } from "../styles/sass.scss";

interface ValidationMessage {
	validationMessage?: string,
	validationType: ValidationType
}

const ValidationMessage = (props: ValidationMessage) => (
	<p 
		/* 
			Normally I would do this kind of styling with classes but I
			wanted to show how I can access sass variables from react.
		*/
		style={{
			color: props.validationType === ValidationType.Success ? 
			green : red
		}}
		className="block full_width margin_top_small">
		{props.validationMessage}
	</p>
)

export default ValidationMessage;