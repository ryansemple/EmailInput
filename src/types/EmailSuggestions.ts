import EmailSuggestion from "./EmailSuggestion";
import { emailDomains, domainExtensions } from "../utility/Email";
import { atCharacter } from "../utility/String";

interface EmailSuggestionsInterface {
	emailSuggestions: EmailSuggestion[],
	returnValidEmailSuggestions: (userEnteredEmail: string) => string[]
}

class EmailSuggestions implements EmailSuggestionsInterface {
	public readonly emailSuggestions: EmailSuggestion[];
	
	constructor()
	{ 
		let emailSuggestions: EmailSuggestion[] = [];
		
		for (let i: number = 0; i < emailDomains.length; i++)
		{
			const emailDomain: string = emailDomains[i];

			for (let i2: number = 0; i2 < domainExtensions.length; i2++)
			{
				const domainExtension: string = domainExtensions[i2];
				
				emailSuggestions.push
				(
					new EmailSuggestion(emailDomain, domainExtension)
				);
			}
		}

		this.emailSuggestions = emailSuggestions;
	}

	public returnValidEmailSuggestions = (userEnteredEmail: string): string[] =>
	{
		if (!userEnteredEmail.includes(atCharacter)) 
		{
			return this.returnAllEmailSuggestions(userEnteredEmail);
		}
		else 
		{
			const emailSplitByAtCharacter: string[] = 
				userEnteredEmail.split(atCharacter);
			const emailCharactersAfterAtCharacter: string = 
				emailSplitByAtCharacter[1];

			if (emailCharactersAfterAtCharacter === "") 
			{
				return this.returnAllEmailSuggestions(userEnteredEmail);
			}
			else
			{
				return this.returnFitleredEmailSuggestions(
					userEnteredEmail, 
					emailCharactersAfterAtCharacter
				);
			}
		}
	}

	private returnAllEmailSuggestions = (userEnteredEmail: string): string[] =>
	{
		let validEmailSuggestions: string[] = [];
		
		for (let i: number = 0; i < this.emailSuggestions.length; i++)
		{
			const currentEmailSuggestion: EmailSuggestion = this.emailSuggestions[i];
			const validEmailSuggestion = currentEmailSuggestion.returnFullEmail(
				userEnteredEmail
			);
			validEmailSuggestions.push(validEmailSuggestion);
		}

		return validEmailSuggestions;
	}

	private returnFitleredEmailSuggestions =
	(userEnteredEmail: string, userEnteredEmailAfterAtCharacter: string) : string[] => 
	{
		let validEmailSuggestions: string[] = [];

		for (let i: number = 0; i < this.emailSuggestions.length; i++)
		{
			const currentEmailSuggestion: EmailSuggestion = this.emailSuggestions[i];
			const currentEmailSuggestionAfterAtCharacter: string =
				currentEmailSuggestion.returnEmailAfterAtCharacter();
			const regexToMatch: RegExp = new RegExp
			(
				`^${userEnteredEmailAfterAtCharacter}`
			);

			const userEmailMatchesSuggestedEmailExactly = 
				userEnteredEmailAfterAtCharacter ===
				currentEmailSuggestionAfterAtCharacter;
			
			if(userEmailMatchesSuggestedEmailExactly) 
			{
				continue;
			}

			if (currentEmailSuggestionAfterAtCharacter.match(regexToMatch))
			{
				const validEmailSuggestion = currentEmailSuggestion.returnFullEmail(
					userEnteredEmail
				);

				validEmailSuggestions.push(validEmailSuggestion);
			}
		}

		return validEmailSuggestions;
	}
}

export default EmailSuggestions;