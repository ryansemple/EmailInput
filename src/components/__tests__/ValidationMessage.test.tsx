import React from "react";
import { render } from "@testing-library/react";
import ValidationMessage from "../ValidationMessage";
import { ValidationType } from "../../types/ValidationType";

test("Renders ValidationMessage component correctly.", () => {
	const { container } = render(
		<ValidationMessage 
			validationMessage="This is valid"
			validationType={ValidationType.Success}
		/>
	);

	expect(container).toMatchSnapshot();
});