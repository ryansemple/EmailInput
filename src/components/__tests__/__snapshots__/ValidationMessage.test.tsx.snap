// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders ValidationMessage component correctly. 1`] = `
<div>
  <p
    class="block full_width margin_top_small bold"
  >
    This is valid
  </p>
</div>
`;
